_posts/2020-09-01-processing-java-stream.md

```diff
@@ -0,0 +1,123 @@
+
+```java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd");
+    public static void main(String[] args) {
+        List<String> a = LIST.stream()
+                .filter(e -> {
+                    System.out.println("Stream - filter: " + e);
+                    return e.startsWith("a");
+                })
+                .map(e -> {
+                    System.out.println("Stream - map:" + e);
+                    return e.toUpperCase();
+                })
+                .collect(Collectors.toList());
+    }
+```
+Z logów wyświetlonych w konsoli widzimy, że operacje wykonywane są horyzontalnie, to znaczy dla każdego elementu od początku do końca. Dzięki temu, sekwencja skończy się na pierwszym niespełnionym warunku.
+```
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: bbb
+Stream - filter: ccc
+Stream - filter: ddd
+```
+
+# Przetwarzanie równoległe
+Wartościowym mechanizmem, który został dodany do Stream API jest możliwość równoległego przetwarzania strumieni, co daje możliwość wykorzystania większej liczby rdzeni procesora do wykonania zadania.
+Jak widać na załączonym kodzie wystarczy, że skorzystamy z metody **parallelStream** zamiast **stream** i nasze dane zostaną odpowiednio podzielone i przetworzone równolegle z użyciem wielu wątków. Warto zwrócić uwagę na to, że samo dzielenie danych wymaga dodatkowej pracy, dlatego przetwarzanie równoległe mniejszej ilości danych może trwać dłużej niż zrobienie tego samego z wykorzystaniem jednego wątku.
+``` java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd");
+    public static void main(String[] args) {
+        LIST.parallelStream()
+                .filter(e -> {
+                    System.out.println("Stream - filter: " + e);
+                    return e.startsWith("a");
+                })
+                .map(e -> {
+                    System.out.println("Stream - map:" + e);
+                    return e.toUpperCase();
+                })
+                .collect(Collectors.toList());
+    }
+```
+
+# Benchmark
+Wiemy już jak przetwarzać strumienie w bardziej efektywnie. Nasuwa się pytanie jak ich szybkość wypada w porównaniu do pętli, aby to sprawdzić zostały przeprowadzone testy przy użyciu Java Microbenchmark Harness i procesora Intel i7 10700K. Kod zaprezentowany niżej został uruchomiony na kolekcjach liczącej 1, 100, 10 000 oraz 1 000 000 elementów.
+
+
+Jednostką pomiarową jest ilość operacji na sekundę, czyli innymi słowy ile razy udało się wywołać daną metodę w ciągu jednej sekundy.
```

Comment:
przecinek:
==> czyli innymi słowy, ile

"udało się" w przypadku wpisu blgoowego nie jest pożądanym zwrotem, bo może sugerować że coś sie Tobie udało, ale czytelnikowi już niekoniecznie. Chyba, że to nie jest pewne, wtedy się nie czepiam :) Niemniej proponuję twarde:
==> ile razy wywołano daną metodę w ciągu jednej sekundy.